Add weasyprint pdf generation as alternative to wkhtmltopdf

bb opened the pull request:
Overview
----------------------------------------

For PDF generation, we currently have the choice between DOMPDF and wkhtmltopdf (via Snappy).

The wkhtmltopdf project itself has been abandoned long time ago as described on their status page: https://wkhtmltopdf.org/status.html#recommendations. An alternative recommended by the wkhtmltopdf maintainer on this status page is [Weasyprint](https://weasyprint.org/). It can be used in PHP via https://github.com/pontedilana/php-weasyprint which has basically the same API as Snappy.

Before
----------------------------------------
If wkhtmltopdf path is configured in the settings, it is used, otherwise PDFDOM.

After
----------------------------------------
if weasyprint path is configured in the settings, it is used. Otherwise (as before): If wkhtmltopdf path is configured in the settings, it is used, otherwise PDFDOM.

Technical Details
----------------------------------------

wkhtmltopdf is mentioned in comments in PDFTrait.php and PDFLetterCommon.php, but I didn't add references to weasyprint there.

Comments
----------------------------------------

PHP version compatibility: Current version of php-weasyprint requires PHP >= 7.4 whereas civicrm-core currently still allowes PHP ~7.3.

Dependencies:

The code for wkhtmltopdf was (and still is) using snappy which is loaded via `require_once 'snappy/src/autoload.php';`, however it is not referenced in compose.json. It could be added e.g. by using `composer require pontedilana/php-weasyprint` but it was not done by those who implemented it initially.

The new code for weasyprint uses [php-weasyprint](https://github.com/pontedilana/php-weasyprint) could be required using `composer require pontedilana/php-weasyprint`, but I chose not to to keep both implementations as similar as possible.

I'm not familiar with PHP dependency management, so I hope this is a good approach. Open for feedback.

bb committed:
Add weasyprint pdf generation as alternative to wkhtmltopdf

civicrm-builder commented:
Can one of the admins verify this patch?

civicrm-builder commented:
Can one of the admins verify this patch?

mlutfy commented:
jenkins, test this please

mlutfy commented:
add to whitelist

mlutfy commented:
@bb Thank you for the PR! Some thoughts:

- weasyprint looks really nice. I looked at the install docs, usage docs, and as you mentioned it's been around for a long time.
- CiviCRM will drop support for PHP 7.3 next month, so if you don't mind waiting a bit, we can then add the dependency in the `composer.json` of this repo, once the RC is branched (I'm not sure if that's in early-December, or early-January).
- and that way there is no need for the `require_once` statement.

I'll also try to deploy on sites that use PDF generation.

aydun commented:
It's good to see PDF generation getting some attention!

If I understand this right, this is a wrapper around weasyprint which is a python application, so using this requires python & weasyprint installed on the Civi server.  As an optional package (like wkhtmlpdf is), I don't think that is a problem but it might prevent this becoming the default pdf generator.    Also, just wondering whether weasyprint could be wrapped as a Civi extension to make it easier to deploy.

wmortada commented on templates/CRM/Admin/Form/Setting/Miscellaneous.tpl:
```diff
@@ -74,6 +74,12 @@
+      <tr class="crm-miscellaneous-form-block-weasyprintPath">
+        <td class="label">{$form.weasyprintPath.label}</td>
+        <td>{$form.weasyprintPath.html}<br />
+          <p class="description">{ts 1="https://weasyprint.org/"}<a href="%1">weasyprint is an alternative utility for generating PDF's</a> which is a successor to the discontinued wkhtmltopdf. Your system administrator will need to download and install this utility, and enter the executable path here.{/ts}</p>
```
Minor typo - no apostrophe needed in 'PDFs'

```suggestion
          <p class="description">{ts 1="https://weasyprint.org/"}<a href="%1">weasyprint is an alternative utility for generating PDFs</a> which is a successor to the discontinued wkhtmltopdf. Your system administrator will need to download and install this utility, and enter the executable path here.{/ts}</p>
```

bb committed:
Harmonize writing PDFs without apostrophe

Co-authored-by: William Mortada <williammortada@thirdsectordesign.org>

bb commented:
@wmortada FYI ffc8b9d

bb commented:
@aydun good points. I'm new to CiviCRM, so I can't decide.
But I have a humble opinion: Personally, I wouldn't repackage Weasyprint itself as extension but just install it with the base system's package manager. The dependency trees are huge, see below.

Regarding the few lines from this PR wrapping the standalone executable weasyprint: I don't think it's worth an extension. It's just a few lines and they're very, very close to wkhtmltopdf. I even considered using the same code and path-option as for wkhtmltopdf with a boolean switch instead of adding a separate path. But I think this way its easier to grasp, and easier to remove one of them later.

So, the only thing which makes sense to me (besides the current approach) is to move _all_ parts of PDF generation to one or more separate extensions, so one could decide between no-plugin→no-PDFs or any of the extensions. It all looks like lots of work for little value.

### Weasyprint and wkhtmltopdf dependencies
Weasyprint pulls many dependencies (~686 MB on Ubuntu 22 LTS, 110 MB on Alpine), even more than wkhtmltopdf (271 MB on Ubuntu. It's not available any more on alpine due to security concerns which lead to this PR in the first place)

Weasy:
```
# apt install weasyprint --no-install-recommends
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  binutils binutils-aarch64-linux-gnu binutils-common cpp cpp-11 dictionaries-common emacsen-common fontconfig g++ g++-11 gcc gcc-11 gcc-11-base gcc-12-base hyphen-en-us libasan6 libatomic1 libbinutils libblas3
  libboost-dev libboost1.74-dev libc-dev-bin libc6 libc6-dev libcc1-0 libcrypt-dev libctf-nobfd0 libctf0 libdatrie1 libexpat1-dev libfribidi0 libgcc-11-dev libgcc-s1 libgdbm-compat4 libgdbm6 libgfortran5 libgomp1
  libgraphite2-3 libharfbuzz0b libhwasan0 libimagequant0 libisl23 libitm1 libjs-jquery libjs-sphinxdoc libjs-underscore liblapack3 liblbfgsb0 liblcms2-2 liblsan0 libmpc3 libmpdec3 libmpfr6 libnsl-dev libopenblas-dev
  libopenblas-pthread-dev libopenblas0 libopenblas0-pthread libopenjp2-7 libpango-1.0-0 libpangoft2-1.0-0 libperl5.34 libpython3-dev libpython3-stdlib libpython3.10 libpython3.10-dev libpython3.10-minimal
  libpython3.10-stdlib libraqm0 libreadline8 libstdc++-11-dev libstdc++6 libtext-iconv-perl libthai-data libthai0 libtirpc-dev libtsan0 libubsan1 libwebpdemux2 libwebpmux3 libxsimd-dev linux-libc-dev mailcap mime-support
  perl perl-base perl-modules-5.34 python-tinycss2-common python3 python3-appdirs python3-attr python3-beniget python3-brotli python3-cffi python3-cffi-backend python3-cssselect2 python3-decorator python3-dev
  python3-distutils python3-fonttools python3-fs python3-gast python3-html5lib python3-lib2to3 python3-lxml python3-lz4 python3-minimal python3-mpmath python3-numpy python3-pil python3-pkg-resources python3-ply
  python3-pycparser python3-pydyf python3-pyphen python3-pythran python3-scipy python3-six python3-sympy python3-tinycss2 python3-tz python3-ufolib2 python3-unicodedata2 python3-webencodings python3.10 python3.10-dev
  python3.10-minimal readline-common rpcsvc-proto shared-mime-info unicode-data zlib1g-dev
Suggested packages:
  binutils-doc cpp-doc gcc-11-locales ispell | aspell | hunspell wordlist gcc-11-doc gcc-multilib make manpages-dev autoconf automake libtool flex bison gdb gcc-doc libreoffice-writer libboost-doc libboost1.74-doc
  libboost-atomic1.74-dev libboost-chrono1.74-dev libboost-container1.74-dev libboost-context1.74-dev libboost-contract1.74-dev libboost-coroutine1.74-dev libboost-date-time1.74-dev libboost-exception1.74-dev
  libboost-fiber1.74-dev libboost-filesystem1.74-dev libboost-graph1.74-dev libboost-graph-parallel1.74-dev libboost-iostreams1.74-dev libboost-locale1.74-dev libboost-log1.74-dev libboost-math1.74-dev libboost-mpi1.74-dev
  libboost-mpi-python1.74-dev libboost-numpy1.74-dev libboost-program-options1.74-dev libboost-python1.74-dev libboost-random1.74-dev libboost-regex1.74-dev libboost-serialization1.74-dev libboost-stacktrace1.74-dev
  libboost-system1.74-dev libboost-test1.74-dev libboost-thread1.74-dev libboost-timer1.74-dev libboost-type-erasure1.74-dev libboost-wave1.74-dev libboost1.74-tools-dev libmpfrc++-dev libntl-dev libboost-nowide1.74-dev
  glibc-doc locales gdbm-l10n liblcms2-utils libstdc++-11-doc libxsimd-doc perl-doc libterm-readline-gnu-perl | libterm-readline-perl-perl libtap-harness-archive-perl python3-doc python3-tk python3-venv python-attr-doc
  python-cssselect2-doc python3-genshi python-lxml-doc python-mpmath-doc python3-gmpy2 python3-matplotlib gfortran python-numpy-doc python3-pytest python-pil-doc python3-setuptools python-ply-doc python-scipy-doc
  texlive-fonts-extra dvipng python-sympy-doc python-tinycss2-doc python3.10-venv python3.10-doc binfmt-support readline-doc
Recommended packages:
  manpages manpages-dev libc-devtools libnss-nis libnss-nisplus javascript-common bzip2 file xz-utils netbase python3-bs4 python3-olefile
The following NEW packages will be installed:
  binutils binutils-aarch64-linux-gnu binutils-common cpp cpp-11 dictionaries-common emacsen-common fontconfig g++ g++-11 gcc gcc-11 gcc-11-base hyphen-en-us libasan6 libatomic1 libbinutils libblas3 libboost-dev
  libboost1.74-dev libc-dev-bin libc6-dev libcc1-0 libcrypt-dev libctf-nobfd0 libctf0 libdatrie1 libexpat1-dev libfribidi0 libgcc-11-dev libgdbm-compat4 libgdbm6 libgfortran5 libgomp1 libgraphite2-3 libharfbuzz0b
  libhwasan0 libimagequant0 libisl23 libitm1 libjs-jquery libjs-sphinxdoc libjs-underscore liblapack3 liblbfgsb0 liblcms2-2 liblsan0 libmpc3 libmpdec3 libmpfr6 libnsl-dev libopenblas-dev libopenblas-pthread-dev
  libopenblas0 libopenblas0-pthread libopenjp2-7 libpango-1.0-0 libpangoft2-1.0-0 libperl5.34 libpython3-dev libpython3-stdlib libpython3.10 libpython3.10-dev libpython3.10-minimal libpython3.10-stdlib libraqm0
  libreadline8 libstdc++-11-dev libtext-iconv-perl libthai-data libthai0 libtirpc-dev libtsan0 libubsan1 libwebpdemux2 libwebpmux3 libxsimd-dev linux-libc-dev mailcap mime-support perl perl-modules-5.34
  python-tinycss2-common python3 python3-appdirs python3-attr python3-beniget python3-brotli python3-cffi python3-cffi-backend python3-cssselect2 python3-decorator python3-dev python3-distutils python3-fonttools python3-fs
  python3-gast python3-html5lib python3-lib2to3 python3-lxml python3-lz4 python3-minimal python3-mpmath python3-numpy python3-pil python3-pkg-resources python3-ply python3-pycparser python3-pydyf python3-pyphen
  python3-pythran python3-scipy python3-six python3-sympy python3-tinycss2 python3-tz python3-ufolib2 python3-unicodedata2 python3-webencodings python3.10 python3.10-dev python3.10-minimal readline-common rpcsvc-proto
  shared-mime-info unicode-data weasyprint zlib1g-dev
The following packages will be upgraded:
  gcc-12-base libc6 libgcc-s1 libstdc++6 perl-base
5 upgraded, 128 newly installed, 0 to remove and 24 not upgraded.
Need to get 146 MB of archives.
After this operation, 686 MB of additional disk space will be used.
```

wkhtmltopdf:
```
apt install wkhtmltopdf --no-install-recommends
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  fontconfig iso-codes libavahi-client3 libavahi-common-data libavahi-common3 libcups2 libdbus-1-3 libdouble-conversion3 libdrm-amdgpu1 libdrm-common libdrm-nouveau2 libdrm-radeon1 libdrm2 libdw1 libegl-mesa0 libegl1
  libevdev2 libgbm1 libgl1 libgl1-mesa-dri libglapi-mesa libglvnd0 libglx-mesa0 libglx0 libgraphite2-3 libgstreamer-plugins-base1.0-0 libgstreamer1.0-0 libgudev-1.0-0 libharfbuzz0b libhyphen0 libice6 libinput-bin
  libinput10 libllvm15 libmd4c0 libmtdev1 liborc-0.4-0 libpcre2-16-0 libqt5core5a libqt5dbus5 libqt5gui5 libqt5network5 libqt5positioning5 libqt5printsupport5 libqt5qml5 libqt5qmlmodels5 libqt5quick5 libqt5sensors5
  libqt5svg5 libqt5webchannel5 libqt5webkit5 libqt5widgets5 libsensors-config libsensors5 libsm6 libunwind8 libwacom-common libwacom9 libwayland-client0 libwayland-server0 libwoff1 libx11-xcb1 libxcb-dri2-0 libxcb-dri3-0
  libxcb-glx0 libxcb-icccm4 libxcb-image0 libxcb-keysyms1 libxcb-present0 libxcb-randr0 libxcb-render-util0 libxcb-render0 libxcb-shape0 libxcb-shm0 libxcb-sync1 libxcb-util1 libxcb-xfixes0 libxcb-xinerama0 libxcb-xinput0
  libxcb-xkb1 libxext6 libxfixes3 libxkbcommon-x11-0 libxkbcommon0 libxrender1 libxshmfence1 libxxf86vm1 shared-mime-info x11-common xkb-data
Suggested packages:
  isoquery cups-common libvisual-0.4-plugins gstreamer1.0-tools libthai0 qt5-image-formats-plugins qtwayland5 qt5-qmltooling-plugins lm-sensors
Recommended packages:
  dbus libgl1-amber-dri gstreamer1.0-plugins-base qttranslations5-l10n qt5-gtk-platformtheme geoclue-2.0 libwacom-bin xserver | xvfb
The following NEW packages will be installed:
  fontconfig iso-codes libavahi-client3 libavahi-common-data libavahi-common3 libcups2 libdbus-1-3 libdouble-conversion3 libdrm-amdgpu1 libdrm-common libdrm-nouveau2 libdrm-radeon1 libdrm2 libdw1 libegl-mesa0 libegl1
  libevdev2 libgbm1 libgl1 libgl1-mesa-dri libglapi-mesa libglvnd0 libglx-mesa0 libglx0 libgraphite2-3 libgstreamer-plugins-base1.0-0 libgstreamer1.0-0 libgudev-1.0-0 libharfbuzz0b libhyphen0 libice6 libinput-bin
  libinput10 libllvm15 libmd4c0 libmtdev1 liborc-0.4-0 libpcre2-16-0 libqt5core5a libqt5dbus5 libqt5gui5 libqt5network5 libqt5positioning5 libqt5printsupport5 libqt5qml5 libqt5qmlmodels5 libqt5quick5 libqt5sensors5
  libqt5svg5 libqt5webchannel5 libqt5webkit5 libqt5widgets5 libsensors-config libsensors5 libsm6 libunwind8 libwacom-common libwacom9 libwayland-client0 libwayland-server0 libwoff1 libx11-xcb1 libxcb-dri2-0 libxcb-dri3-0
  libxcb-glx0 libxcb-icccm4 libxcb-image0 libxcb-keysyms1 libxcb-present0 libxcb-randr0 libxcb-render-util0 libxcb-render0 libxcb-shape0 libxcb-shm0 libxcb-sync1 libxcb-util1 libxcb-xfixes0 libxcb-xinerama0 libxcb-xinput0
  libxcb-xkb1 libxext6 libxfixes3 libxkbcommon-x11-0 libxkbcommon0 libxrender1 libxshmfence1 libxxf86vm1 shared-mime-info wkhtmltopdf x11-common xkb-data
0 upgraded, 91 newly installed, 0 to remove and 29 not upgraded.
Need to get 66.3 MB of archives.
After this operation, 271 MB of additional disk space will be used.
```

weasyprint on alpine:
```
docker run --rm -it alpine
/ # apk update
fetch https://dl-cdn.alpinelinux.org/alpine/v3.18/main/aarch64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.18/community/aarch64/APKINDEX.tar.gz
v3.18.4-239-gcdb9428c46f [https://dl-cdn.alpinelinux.org/alpine/v3.18/main]
v3.18.4-244-g117d218742b [https://dl-cdn.alpinelinux.org/alpine/v3.18/community]
OK: 19953 distinct packages available
/ # apk stats
installed:
  packages: 15
  dirs: 101
  files: 110
  bytes: 8544256
  triggers: 1
available:
  names: 44759
  packages: 19953
atoms:
  num: 13574
/ # apk add weasyprint
(1/77) Installing libxau (1.0.11-r2)
(2/77) Installing libmd (1.0.4-r2)
(3/77) Installing libbsd (0.11.7-r1)
(4/77) Installing libxdmcp (1.1.4-r2)
(5/77) Installing libxcb (1.15-r1)
(6/77) Installing libx11 (1.8.7-r0)
(7/77) Installing libxrender (0.9.11-r3)
(8/77) Installing libexpat (2.5.0-r1)
(9/77) Installing brotli-libs (1.0.9-r14)
(10/77) Installing libbz2 (1.0.8-r5)
(11/77) Installing libpng (1.6.39-r3)
(12/77) Installing freetype (2.13.0-r5)
(13/77) Installing fontconfig (2.14.2-r3)
(14/77) Installing libxft (2.3.8-r1)
(15/77) Installing libxext (1.3.5-r2)
(16/77) Installing pixman (0.42.2-r1)
(17/77) Installing cairo (1.17.8-r1)
(18/77) Installing fribidi (1.0.13-r0)
(19/77) Installing libffi (3.4.4-r2)
(20/77) Installing libintl (0.21.1-r7)
(21/77) Installing libblkid (2.38.1-r8)
(22/77) Installing libmount (2.38.1-r8)
(23/77) Installing pcre2 (10.42-r1)
(24/77) Installing glib (2.76.4-r0)
(25/77) Installing graphite2 (1.3.14-r5)
(26/77) Installing harfbuzz (7.3.0-r0)
(27/77) Installing pango (1.50.14-r1)
(28/77) Installing gdbm (1.23-r1)
(29/77) Installing xz-libs (5.4.3-r0)
(30/77) Installing libgcc (12.2.1_git20220924-r10)
(31/77) Installing libstdc++ (12.2.1_git20220924-r10)
(32/77) Installing mpdecimal (2.5.1-r2)
(33/77) Installing ncurses-terminfo-base (6.4_p20230506-r0)
(34/77) Installing libncursesw (6.4_p20230506-r0)
(35/77) Installing libpanelw (6.4_p20230506-r0)
(36/77) Installing readline (8.2.1-r1)
(37/77) Installing sqlite-libs (3.41.2-r2)
(38/77) Installing python3 (3.11.6-r0)
(39/77) Installing python3-pycache-pyc0 (3.11.6-r0)
(40/77) Installing pyc (0.1-r0)
(41/77) Installing py3-webencodings (0.5.1-r7)
(42/77) Installing py3-webencodings-pyc (0.5.1-r7)
(43/77) Installing py3-tinycss2 (1.2.1-r3)
(44/77) Installing py3-tinycss2-pyc (1.2.1-r3)
(45/77) Installing py3-cssselect2 (0.7.0-r3)
(46/77) Installing py3-cssselect2-pyc (0.7.0-r3)
(47/77) Installing py3-cparser (2.21-r2)
(48/77) Installing py3-cparser-pyc (2.21-r2)
(49/77) Installing py3-cffi (1.15.1-r3)
(50/77) Installing py3-cffi-pyc (1.15.1-r3)
(51/77) Installing py3-fonttools (4.39.4-r0)
(52/77) Installing py3-fonttools-pyc (4.39.4-r0)
(53/77) Installing py3-six (1.16.0-r6)
(54/77) Installing py3-six-pyc (1.16.0-r6)
(55/77) Installing py3-html5lib (1.1-r4)
(56/77) Installing py3-html5lib-pyc (1.1-r4)
(57/77) Installing libimagequant (4.2.0-r0)
(58/77) Installing libjpeg-turbo (2.1.5.1-r3)
(59/77) Installing lcms2 (2.15-r2)
(60/77) Installing openjpeg (2.5.0-r3)
(61/77) Installing libwebp (1.3.2-r0)
(62/77) Installing zstd-libs (1.5.5-r4)
(63/77) Installing tiff (4.5.1-r0)
(64/77) Installing py3-pillow (9.5.0-r1)
(65/77) Installing py3-pillow-pyc (9.5.0-r1)
(66/77) Installing py3-pydyf (0.6.0-r1)
(67/77) Installing py3-pydyf-pyc (0.6.0-r1)
(68/77) Installing py3-pyphen (0.13.2-r1)
(69/77) Installing py3-pyphen-pyc (0.13.2-r1)
(70/77) Installing zopfli (1.0.3-r1)
(71/77) Installing py3-zopfli (0.2.2-r2)
(72/77) Installing py3-zopfli-pyc (0.2.2-r2)
(73/77) Installing weasyprint-pyc (58.1-r0)
(74/77) Installing py3-brotli-pyc (1.0.9-r14)
(75/77) Installing python3-pyc (3.11.6-r0)
(76/77) Installing py3-brotli (1.0.9-r14)
(77/77) Installing weasyprint (58.1-r0)
Executing busybox-1.36.1-r2.trigger
Executing glib-2.76.4-r0.trigger
OK: 110 MiB in 92 packages
/ # apk stats
installed:
  packages: 92
  dirs: 401
  files: 3258
  bytes: 116297728
  triggers: 4
available:
  names: 44762
  packages: 19953
atoms:
```

bb commented:
@mlutfy you decide if we wait until PHP 7.3 is dropped and we can add php-weasyprint to composer or merge now and add another PR later.
I think another PR later would be fine. This second PR could also add Snappy (for wkhtmltopdf), so basically both extended PDF solutions would be cleaner.

### Failing Tests
@ All: I could need some help with the failing tests. The only occurrence of wkhtml2pdf where I didn't add weasyprint is in MagicMerge.php in `getPropertyMap`. According to the doc comment above, this is just for legacy handling.
Should I add weasyprintPath there as well?
Any other places I'd need to look for hints how to fix the tests?

aydun commented:
@bb re packaging: I agree that if you can use your system's package manager then that's easy.  We certainly don't want to be throwing half an Ubuntu OS into an extension :-) 

I'm thinking about those on shared hosting who can't install system packages but can install Civi extensions.  If they have a functioning python3 installation as commonly provided, what else is needed to run weasyprint?  It looks like we could put the weasyprint python files from github into an extension and use that. I'm no python expert but following the instructions at https://github.com/Kozea/WeasyPrint/blob/main/docs/contribute.rst creates a `venv/bin/weasyprint` that runs and produced sensible output on the (1!) test file I tried - and does not require any additional system packages to be installed.

demeritcowboy commented:
@bb On the tests: There are lots of places in and outside core that use $config or CRM_Core_Config::singleton() so it's maybe deprecated in theory but not in practice. However the easy fix here would be whereever you've added those to get weasyprintPath, use `\Civi::settings()->get('weasyprintPath')` instead.

bb commented:
@aydun I don't know what is typically included in today's shared hosting, I haven't used that for more than a decade. I'm always on my own servers. Did you do your test on shared hosting or on your development machine?
I *guess* the part harder than having Python itself is getting support for Python Wheels (precompiled packages). If Wheels are not supported, you'd need to compile yourself. The non-Python stuff would be similar, e.g. i.e. [pango](https://pango.gnome.org/) depends on glib, cairo, even some X11 libraries. -- I would not expect those to be available on shared hosting.

I guess you've already seen it when you dug up the Weasyprint contributing guide: next to it is the [first steps guide](https://github.com/Kozea/WeasyPrint/blob/main/docs/first_steps.rst) which outlines the direct dependencies for several Linux distributions. If those libs are either statically linked or matching the shared hosting's system, that might work. It also shows that if you don't have support for wheels, you need gcc which I've never seen available on shared hosting.

For me personally, this is not something I'd go into any further, so feel free to give it a shot. I'm sure all the shared hosting users will be happy about great-looking PDFs. Providing just the Weasyprint binary and its dependencies is completely decoupled from this PR and CiviCRM itself, it doesn't even need to be a CiviCRM extension.

As hinted above, I'll only deploy via Docker, most likely via images I build myself.

By the way, you'll probably also want to have a selection of fonts available on the machine. Weasyprint will embed them into the PDFs as needed.

bb committed:
Convert PDF Utils to new config syntax

Partially revert f144ac104a

bb committed:
Change setting name to new settings style

bb committed:
Add contributor

bb commented:
@demeritcowboy Thank you, that helped a lot! I first added weasyprint to MagicMerge which also solved it. Then I went with your proposal and converted the old style to the new style, not only for my changes but for everything inside PDF/Utils.php.
This led to another test failure: https://test.civicrm.org/job/CiviCRM-Core-Matrix-PR/6656/BKPROF=dfl,SUITES=phpunit-crm,label=bknix-tmp/testReport/junit/(root)/CRM_Contribute_Form_Task_PDFLetterCommonTest/testAllContributionTokens/

```diff
-    <style type="text/css">@import url(http://build-0.127.0.0.1.nip.io:5500/css/print.css);</style>\n
+    <style type="text/css">@import url([civicrm.root]/css/print.css);</style>\n
```

Thus, I reverted the change to the new setting style for this single line. But I wonder: What would be the correct way to read this URL using the new style?

I also reverted the MagicMerge part again, because it was no longer needed.

The only remaining failing test after was the change in settings style from camelCase to under_score so I also did that.

Now I hopefully deserved adding myself to the contributions-key file 🚀

demeritcowboy commented:
> But I wonder: What would be the correct way to read this URL using the new style

For `$config->userFrameworkResourceURL` it would be something like `\Civi::Paths()->getUrl(\Civi::settings()->get('userFrameworkResourceURL'))`, but I don't think I've ever seen that way of doing it - most places would just use `$config->userFrameworkResourceURL`. So ¯\_(ツ)_/¯

> Now I hopefully deserved adding myself to the contributions-key file

👍

demeritcowboy commented:
FYI it's not necessary until the end but the way to clean up all the intermediate work is documented at https://docs.civicrm.org/dev/en/latest/tools/git/#rebasing-interactive. It makes sense to have multiple commits for logically separate things, but e.g. the revert could be "squashed" down.

bb commented:
@demeritcowboy I rebased onto upstream/master and squashed the revert-commits 👍

mlutfy commented:
(question for mergers) Should we wait for PHP 7.3 to be dropped when the next RC is branched, so that the library can be added to composer.json?

There is a `require_once 'php-weasyprint/src/autoload.php';` and the lib is not part of this PR.

PR otherwise looks good to me.

bb commented:
To keep the wkhtmltopdf functionality on par with that proposal, we might also want to add snappy to composer.json and remove `require_once 'snappy/src/autoload.php';`  in https://github.com/civicrm/civicrm-core/pull/28273/files#diff-c124040698e738317396a1f24adbe163e6871de7a5bb969ca5b68e941619a199L154

Should we do that? And if so: in this PR or should we have that separate (e.g. for being able to revert separately)?

mlutfy commented:
@bb We can finally merge this now that PHP 7.3 support has been dropped. There is just some tiny details:

Can you add the library to composer:

```
composer require pontedilana/php-weasyprint
```

and then `git add composer.json composer.lock`  to your PR.

and remove the line with `require_once` in `CRM/Utils/PDF/Utils.php`

I would also tweak the strings for the path error messages, but I can also do it in a subsequent PR if you prefer.

Here is a small diff:

```
diff --git a/CRM/Admin/Form/Setting/Miscellaneous.php b/CRM/Admin/Form/Setting/Miscellaneous.php
index ed503dcd8d..76c80cf4f2 100644
--- a/CRM/Admin/Form/Setting/Miscellaneous.php
+++ b/CRM/Admin/Form/Setting/Miscellaneous.php
@@ -130,7 +130,7 @@ class CRM_Admin_Form_Setting_Miscellaneous extends CRM_Admin_Form_Setting {
         !file_exists($path) ||
         !is_executable($path)
       ) {
-        $errors['weasyprint_path'] = ts('The weasyprint_path does not exist or is not valid');
+        $errors['weasyprint_path'] = ts('The path for %1 does not exist or is not valid', [1 => 'weasyprint']);
       }
     }
     if (!empty($fields['wkhtmltopdfPath'])) {
@@ -144,7 +144,7 @@ class CRM_Admin_Form_Setting_Miscellaneous extends CRM_Admin_Form_Setting {
         !file_exists($path) ||
         !is_executable($path)
       ) {
-        $errors['wkhtmltopdfPath'] = ts('The wkhtmltopdfPath does not exist or is not valid');
+        $errors['wkhtmltopdfPath'] = ts('The path for %1 does not exist or is not valid', [1 => 'wkhtmltopdf']);
       }
     }
     return $errors;
diff --git a/CRM/Utils/PDF/Utils.php b/CRM/Utils/PDF/Utils.php
index 5b5f06fc73..53cdf3b0ed 100644
--- a/CRM/Utils/PDF/Utils.php
+++ b/CRM/Utils/PDF/Utils.php
@@ -154,7 +154,6 @@ class CRM_Utils_PDF_Utils {
    * @param string $fileName
    */
   public static function _html2pdf_weasyprint($paper_size, $orientation, $margins, $html, $output, $fileName) {
-    require_once 'php-weasyprint/src/autoload.php';
     $weasyprint = new Pontedilana\PhpWeasyPrint\Pdf(\Civi::settings()->get('weasyprint_path'));
     $weasyprint->setOption("page-width", $paper_size[2] . "pt");
     $weasyprint->setOption("page-height", $paper_size[3] . "pt");
```

(excludes the changes to composer)

bb committed:
Require weasyprint with composer

bb commented:
@mlutfy thanks you!

It‘s all done now, rebased and the tests are green.

For now, I left the require for snappy in and didn’t add snappy in composer. 
I can remove that directly afterwards, together with the corresponding change in the packages repo.

bb commented:
FYI: See https://github.com/civicrm/civicrm-core/pull/28673 and https://github.com/civicrm/civicrm-packages/pull/375.

demeritcowboy commented:
What version of weasyprint is this compatible with? If I try 60.2 I immediately get a crash with unknown option page-width. At https://github.com/pontedilana/php-weasyprint#pass-options-to-phpweasyprint it says you can see the options with `weasyprint -h` and I don't see page-width.

bb commented:
Oh snap... that's my fault. The patch here is different from what I'm actually running and this is backfiring now.

In my installation, I just modified the wkhtmltopdf method without the additional configuration setting. I replaced the snappy code with weasyprint, some debugging code and went all-in with CSS `@page { size: ... }` hard coded (so my installation doesn't even support wkhtmltopdf at the moment).
After that I decided to contribute here, figured that I can't just break the existing code and contributed from what I thought would be compatible but never actually tested in this exact version (besides the configuration part in the test-installation provided by CI here). 

Here's what I need to do now:
* remove all of the `setOption` calls, none of those are needed
* migrate the CSS injection (`@page { size: ... }`) from the my modified wkhtml-method in my installation to the `@page { size: ... }`  to `html2pdf` at the top of the file (where the margins are already injected).
* simplify the method signature because some of the parameters will no longer be needed then
* test that it's actually working 😇

bb committed:
Change Weasyprint to use css page styles

demeritcowboy commented on CRM/Utils/PDF/Utils.php:
```diff
@@ -72,7 +88,7 @@ public static function html2pdf($text, $fileName = 'civicrm.pdf', $output = FALS
   <head>
     <meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-8\"/>
-    <style>@page { margin: {$t}{$metric} {$r}{$metric} {$b}{$metric} {$l}{$metric}; }</style>
+    <style>@page { margin: {$t}{$metric} {$r}{$metric} {$b}{$metric} {$l}{$metric}; $css_page_size }</style>
```
The test fail in CRM_Contribute_Form_Task_PDFLetterCommonTest::testAllContributionTokens is because it's not expecting the extra blank space here. Maybe just include a space at the beginning in line 69 above and then no space here?

The other test I'm not sure, might just be a timeout weirdness?

Speaking of line 69, I think css itself supports the possible values that $paperSize['metric'] might be (in, cm, mm, pt, px, etc) but I guess the question is whether weasyprint does. If it doesn't then one alternate is use the $paper_height, $paper_width that's already been converted to pt and always use pt.

bb replied:
Practically I went with `a4 {$orientation}` until today.  However, with all those special Avery formats for the labels... I didn't want to make a whitelist and fall back to specific units. That'd be yet another can of worms.

Regarding support in theory:
Length Units: https://doc.courtbouillon.org/weasyprint/stable/api_reference.html#css-values-and-units-module-level-3
> Font-related lengths (em, ex, ch, rem), absolute lengths (cm, mm, q, in, pt, pc, px), angles (rad, grad, turn, deg), resolutions (dpi, dpcm, dppx) are supported.

Page sizes: https://doc.courtbouillon.org/weasyprint/stable/api_reference.html#css → https://drafts.csswg.org/css-page-3/#page-size-prop

demeritcowboy commented on CRM/Utils/PDF/Utils.php:
```diff
@@ -72,7 +88,7 @@ public static function html2pdf($text, $fileName = 'civicrm.pdf', $output = FALS
   <head>
     <meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-8\"/>
-    <style>@page { margin: {$t}{$metric} {$r}{$metric} {$b}{$metric} {$l}{$metric}; }</style>
+    <style>@page { margin: {$t}{$metric} {$r}{$metric} {$b}{$metric} {$l}{$metric}; $css_page_size}</style>
```
Sorry not the ending space, the one in front of $css_page_size.

bb replied:
Thanks for your patience. I didn't read your previous comment closely enough last night.

bb commented:
For reference, there's some info about css units support in this resolved conversation: https://github.com/civicrm/civicrm-core/pull/28273#discussion_r1430776076

demeritcowboy commented:
Thanks for sticking with it.
And it does seem to handle fonts/unicode well.
Let's do it.

bb commented:
Thank you all for helping me along the way, especially @demeritcowboy 🤠  and @mlutfy!
